### PR TITLE
GEODE-4080: Protobuf JSON objects are in a proto string

### PR DESCRIPTION
**Pull request by galen-pivotal**
Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [x] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [x] Is your initial contribution a single, squashed commit?

- [x] Does `gradlew build` run cleanly?

- [x] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

**Comment by galen-pivotal** [comment]
@WireBaron @PivotalSarge

File: geode-client-protocol/src/main/java/org/apache/geode/internal/protocol/serialization/TypeConverter.java [hunk 1]
```diff
@@ -24,10 +24,10 @@
  * @param <T> the type this codec knows how to convert
  */
 @Experimental
-public interface TypeCodec<T> {
-  T decode(byte[] incoming);
+public interface TypeConverter<F, T> {
+  T decode(F incoming);
```
Comment (WireBaron) [inline geode-client-protocol/src/main/java/org/apache/geode/internal/protocol/serialization/TypeConverter.java]: Should this have "throws EncodingException"?  (same for encode)

File: geode-protobuf/src/main/java/org/apache/geode/internal/protocol/protobuf/v1/ProtobufSerializationService.java [hunk 2]
```diff
@@ -14,42 +14,124 @@
...
+          builder.setBooleanResult((Boolean) value);
+          break;
+        }
+        case STRING: {
+          builder.setStringResult((String) value);
+          break;
+        }
+      }
+    } catch (UnknownProtobufPrimitiveType unknownProtobufPrimitiveType) {
+      if (value instanceof PdxInstance) {
```
Comment (upthewaterspout) [inline geode-protobuf/src/main/java/org/apache/geode/internal/protocol/protobuf/v1/ProtobufSerializationService.java]: We shouldn't be throwing and catching exceptions for something which is a normal event - in this case handling a PDX value.

**Comment by galen-pivotal** [comment]
Updated in response to feedback and rebased.

File: geode-protobuf/src/main/java/org/apache/geode/internal/protocol/protobuf/v1/utilities/ProtobufEncodingTypes.java [hunk 3]
```diff
@@ -28,22 +35,25 @@
   BOOLEAN(Boolean.class),
   DOUBLE(Double.class),
   FLOAT(Float.class),
-  BINARY(byte[].class);
+  BINARY(byte[].class),
+
+  // This will probably have to change once the protocol supports multiple object encodings.
+  PDX_OBJECT(PdxInstance.class);
```
Comment (WireBaron) [inline geode-protobuf/src/main/java/org/apache/geode/internal/protocol/protobuf/v1/utilities/ProtobufEncodingTypes.java]: I kind of liked having this enum as a way to identify which types were en/de-codable directly by the protobuf code as opposed to those needing a custom encoder/decoder provided by the server (and also by the client).  Without that distinction, does it make sense to move this enum into the ProtobufSerializationService class?

File: geode-client-protocol/src/main/java/org/apache/geode/internal/protocol/serialization/codec/JsonPdxConverter.java [hunk 4]
```diff
@@ -17,19 +17,29 @@
 import org.apache.geode.annotations.Experimental;
```
Comment (WireBaron) [inline geode-client-protocol/src/main/java/org/apache/geode/internal/protocol/serialization/codec/JsonPdxConverter.java]: Hmm, curiously I can't seem to add a comment to line 15, but did we want to rethink the package name now that we're no longer calling these codecs?

**Pull request merged**